docs/man/man1/reagent.1
.\" Automatically generated by Pod::Man 2.23 (Pod::Simple 3.16)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.ie \nF \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
..
.    nr % 0
.    rr F
.\}
.el \{\
.    de IX
..
.\}
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "REAGENT 1"
.TH REAGENT 1 "2011-08-11" "perl v5.12.3" "Entropy"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
reagent \- Official Sabayon Linux Server\-Side Package Management tool
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
\&\fBreagent\fR [action] [basic_options] [options] [atom | package_file | \f(CW@set\fR] ...
.PP
\&\fBreagent\fR [action] [basic_options] [options] ...
.PP
\&\fBreagent\fR \-\-help
.PP
\&\fBreagent\fR \-\-version
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
\&\fBreagent\fR is a server-side repository administration tool used to sync Portage
database with Entropy by phisically creating package files and updating Entropy
repository database.
.SH "BASIC_OPTIONS"
.IX Header "BASIC_OPTIONS"
.IP "\fB\-\-help (\-h)\fR" 4
.IX Item "--help (-h)"
this output
.IP "\fB\-\-version\fR" 4
.IX Item "--version"
print version
.IP "\fB\-\-nocolor\fR" 4
.IX Item "--nocolor"
disable colorized output
.IP "\fB\-\-color\fR" 4
.IX Item "--color"
force colorized output
.SH "OPTIONS"
.IX Header "OPTIONS"
.IP "\fBstatus\fR" 4
.IX Item "status"
show current repositories status
.IP "\fBupdate\fR" 4
.IX Item "update"
scan the System looking for newly compiled packages
.RS 4
.IP "\fB\-\-seekstore\fR" 4
.IX Item "--seekstore"
analyze the Entropy Store directory directly
.IP "\fB\-\-repackage <atoms\fR>" 4
.IX Item "--repackage <atoms>"
repackage the specified atoms
.IP "\fB\-\-noask\fR" 4
.IX Item "--noask"
do not ask anything except critical things
.IP "\fB\-\-atoms <atoms\fR>" 4
.IX Item "--atoms <atoms>"
manage only the specified atoms
.IP "\fB\-\-interactive\fR" 4
.IX Item "--interactive"
run in interactive mode (asking things one by one)
.RE
.RS 4
.RE
.IP "\fBinject <packages\fR>" 4
.IX Item "inject <packages>"
add binary packages to repository w/o affecting scopes (multipackages)
.IP "\fBquery\fR" 4
.IX Item "query"
do some searches into repository databases
.RS 4
.IP "\fBsearch\fR" 4
.IX Item "search"
search packages inside the default repository database
.IP "\fBmatch\fR" 4
.IX Item "match"
match package dependency inside the default repository database
.IP "\fBneeded\fR" 4
.IX Item "needed"
show runtime libraries needed by the provided atoms
.IP "\fBtags\fR" 4
.IX Item "tags"
show packages owning the specified tags
.IP "\fBsets\fR" 4
.IX Item "sets"
search available package sets
.IP "\fBfiles\fR" 4
.IX Item "files"
show files owned by the provided atoms
.IP "\fBbelongs\fR" 4
.IX Item "belongs"
show from what package the provided files belong
.IP "\fBdescription\fR" 4
.IX Item "description"
search packages by description
.IP "\fBlist\fR" 4
.IX Item "list"
list all the packages in the default repository
.IP "\fBgraph\fR" 4
.IX Item "graph"
show direct depdendencies tree for provided installable atoms
.RS 4
.IP "\fB\-\-complete\fR" 4
.IX Item "--complete"
include system packages, build deps and circularity information
.RE
.RS 4
.RE
.IP "\fBrevgraph\fR" 4
.IX Item "revgraph"
show reverse depdendencies tree for provided installed atoms
.RS 4
.IP "\fB\-\-complete\fR" 4
.IX Item "--complete"
include system packages, build deps and circularity information
.RE
.RS 4
.RE
.IP "\fB\-\-verbose\fR" 4
.IX Item "--verbose"
show more details
.IP "\fB\-\-quiet\fR" 4
.IX Item "--quiet"
print results in a scriptable way
.RE
.RS 4
.RE
.IP "\fBrepo\fR" 4
.IX Item "repo"
manage a repository
.RS 4
.IP "\fBinit\fR" 4
.IX Item "init"
(re)initialize the current repository
.IP "\fBbump\fR" 4
.IX Item "bump"
manually force a revision bump for the current repository database
.RS 4
.IP "\fB\-\-sync\fR" 4
.IX Item "--sync"
synchronize the database
.RE
.RS 4
.RE
.IP "\fBflushback [branches]\fR" 4
.IX Item "flushback [branches]"
flush back old branches packages to current branch
.IP "\fBremove\fR" 4
.IX Item "remove"
remove the provided atoms from the current repository database
.RS 4
.IP "\fB\-\-nodeps\fR" 4
.IX Item "--nodeps"
do not include reverse dependencies
.RE
.RS 4
.RE
.IP "\fBmultiremove\fR" 4
.IX Item "multiremove"
remove the provided injected atoms (all if no atom specified)
.IP "\fBcreate-empty-database\fR" 4
.IX Item "create-empty-database"
create an empty repository database in the provided path
.IP "\fBswitchbranch <from branch\fR <to branch>>" 4
.IX Item "switchbranch <from branch <to branch>>"
switch to the specified branch the repository
.IP "\fBmd5remote [atoms]\fR" 4
.IX Item "md5remote [atoms]"
verify remote integrity of the provided atoms
.IP "\fBbackup\fR" 4
.IX Item "backup"
backup current repository database
.IP "\fBrestore\fR" 4
.IX Item "restore"
restore a previously backed-up repository database
.IP "\fBenable <repo\fR>" 4
.IX Item "enable <repo>"
enable the specified repository
.IP "\fBdisable <repo\fR>" 4
.IX Item "disable <repo>"
disable the specified repository
.IP "\fBpackage-dep-check <repo\fR [atoms]>" 4
.IX Item "package-dep-check <repo [atoms]>"
check packages in repository for missing dependencies
.IP "\fBpackage-dep <repo\fR [atoms]>" 4
.IX Item "package-dep <repo [atoms]>"
handle packages dependencies
.IP "\fBpackage-tag <repo\fR <tag\-string> [atoms]>" 4
.IX Item "package-tag <repo <tag-string> [atoms]>"
clone a package assigning it an arbitrary tag
.IP "\fBpackage-mask <repo\fR [atoms]>" 4
.IX Item "package-mask <repo [atoms]>"
mask given package in given repository
.IP "\fBpackage-unmask <repo\fR [atoms]>" 4
.IX Item "package-unmask <repo [atoms]>"
unmask given package in given repository
.IP "\fBmove <from\fR <to> [atoms]>" 4
.IX Item "move <from <to> [atoms]>"
move packages from a repository to another
.IP "\fBcopy <from\fR <to> [atoms]>" 4
.IX Item "copy <from <to> [atoms]>"
copy packages from a repository to another
.IP "\fBdefault <repo_id\fR>" 4
.IX Item "default <repo_id>"
set the default repository
.RE
.RS 4
.RE
.IP "\fBkey\fR" 4
.IX Item "key"
manage repository digital signatures (OpenGPG)
.RS 4
.IP "\fBcreate [repos]\fR" 4
.IX Item "create [repos]"
create keypair for repositories and sign packages
.IP "\fBdelete [repos]\fR" 4
.IX Item "delete [repos]"
delete keypair (and digital signatures) of repository
.IP "\fBstatus [repos]\fR" 4
.IX Item "status [repos]"
show currently configured keys information for given repositories
.IP "\fBsign [repos]\fR" 4
.IX Item "sign [repos]"
sign (or re-sign) packages in repository using currently set keypair
.IP "\fBimport <repo_id\fR <privkey_path> <pubkey_path>>" 4
.IX Item "import <repo_id <privkey_path> <pubkey_path>>"
import keypair, bind to given repository
.IP "\fBexport-public <repo_id\fR <key_path>>" 4
.IX Item "export-public <repo_id <key_path>>"
export public key of given repository
.RE
.RS 4
.RE
.IP "\fBexport-private <repo_id\fR <key_path>>" 4
.IX Item "export-private <repo_id <key_path>>"
export private key of given repository
.IP "\fBspm\fR" 4
.IX Item "spm"
source package manager functions
.RS 4
.IP "\fBcompile\fR" 4
.IX Item "compile"
compilation function
.RS 4
.IP "\fBcategories\fR" 4
.IX Item "categories"
compile packages belonging to the provided categories
.RS 4
.IP "\fB\-\-list\fR" 4
.IX Item "--list"
just list packages
.RE
.RS 4
.RE
.IP "\fBpkgset\fR" 4
.IX Item "pkgset"
compile packages in provided package set names
.RS 4
.IP "\fB\-\-list\fR" 4
.IX Item "--list"
just list packages
.IP "\fB\-\-rebuild\fR" 4
.IX Item "--rebuild"
rebuild everything
.IP "\fB\-\-dbupdate\fR" 4
.IX Item "--dbupdate"
run database update if all went fine
.IP "\fB\-\-dbsync\fR" 4
.IX Item "--dbsync"
run mirror sync if all went fine
.RE
.RS 4
.RE
.RE
.RS 4
.RE
.IP "\fBorphans\fR" 4
.IX Item "orphans"
scan orphaned packages in \s-1SPM\s0

docs/ChangeLog
commit 266fc224abe3f00312bcab4a878a01a391bd41c9
Author: Fabio Erculiani <lxnay@sabayon.org>
Date:   Thu Aug 25 19:36:52 2011 +0200

    Release Entropy 1.0_rc44

commit 6059629452d2898b4f66b5a632f3d7834fe05d10
Author: Fabio Erculiani <lxnay@sabayon.org>
Date:   Thu Aug 25 19:36:00 2011 +0200

    [entropy.spm] PortagePlugin.add_installed_package: chmod vardb entry to 755 as Portage does

commit 949f2ab3f3ed1c844a02670fe11b65fb6a284031
Author: Fabio Erculiani <lxnay@sabayon.org>
Date:   Thu Aug 25 18:39:04 2011 +0200

    [entropy.server] add PKG_KEYWORDS variable to those exposed to metadata qa check hook

commit b0a2bbf4fe6d453a63b35be3b17104a8121ccefd
Author: Fabio Erculiani <lxnay@sabayon.org>
Date:   Thu Aug 25 09:52:42 2011 +0200

    [entropy.server.mirrors] do not handle .asc files anymore in tidy_mirrors()

commit 7ead04fa0180e0d79a66804ecd6c162876a72183
Author: Fabio Erculiani <lxnay@sabayon.org>
Date:   Thu Aug 25 09:52:17 2011 +0200

    [kernel-switcher] catch KeyboardInterrupt and don't throw traceback to user

commit 1d29a15e056947066157571d8e4d7f37c0415279
Author: Fabio Erculiani <lxnay@sabayon.org>
Date:   Wed Aug 24 13:45:11 2011 +0200

    [matter] remove temp file in buildfail-tinderbox.sh

commit 2fca78342580aa3897b7b217f9ae5572f2d8ecf2
Author: Fabio Erculiani <lxnay@sabayon.org>
Date:   Wed Aug 24 13:43:35 2011 +0200

    [matter] add simple buildfail example script

commit 7b5d009149aa65aa92d2351f9e6f65fa09d6ad10
Author: Fabio Erculiani <lxnay@sabayon.org>
Date:   Wed Aug 24 12:32:35 2011 +0200

    [matter] add bugzilla submit buildfail hook

commit 3a9cbb6341e55ab803c86523770241861a16e5dc
Author: Fabio Erculiani <lxnay@sabayon.org>
Date:   Wed Aug 24 09:57:58 2011 +0200

    [matter] add "buildfail" directive support

commit 9b22ff56136593fed1e3e5d687c2ac157c44140f
Author: Fabio Erculiani <lxnay@sabayon.org>
Date:   Wed Aug 24 08:48:29 2011 +0200

    [matter] add support for "not-installed" directive

commit 32c6dc7a011d3c258a2d8586d41e35ec061c3abe
Author: Fabio Erculiani <lxnay@sabayon.org>
Date:   Wed Aug 24 08:25:43 2011 +0200

    [matter] add FEATURES="split-log" and CMAKE_NO_COLOR="yes" to builder env

commit ca3fa9daba03760cfedee54ce8b3cb2955605124
Author: Fabio Erculiani <lxnay@sabayon.org>
Date:   Sun Aug 21 14:34:23 2011 +0200

    [entropy] add .gitattributes file, exclude test suite from archive for now (takes too much space)

commit 698129ba8460e7d27d6662c6889387404ee66c25
Author: Fabio Erculiani <lxnay@sabayon.org>
Date:   Sun Aug 21 14:23:04 2011 +0200

    Tagging Entropy version 1.0_rc43

commit 9366621d90e600874cfa3c1a284be72503cecd11
Author: Fabio Erculiani <lxnay@sabayon.org>
Date:   Sun Aug 21 14:23:02 2011 +0200

    Release Entropy 1.0_rc43

commit 13c3b7513d5c2664230e5effe8be8522635d7652
Author: Fabio Erculiani <lxnay@sabayon.org>
Date:   Sun Aug 21 14:17:48 2011 +0200

    [entropy.server.mirrors] implement herustic able to use remote file copy when possible
    
    When dealing with multiple repositories, it is possible that a package file
    gets uploaded multiple times when moved from a repository to another.
    In this scenario, there is a waste of upload bandwidth. This commit
    implements a simple herustic that is able to determine if it's possible
    to copy the file remotely instead of uploading it, if and only if:
    
    - package file with the same name is already available on another
      tracked repository sharing the same mirror URI
    - local package and remote package have the same md5 hash

commit a61f6b99cc567b3aec8a9db36f8e454e6e34cd47
Author: Fabio Erculiani <lxnay@sabayon.org>
Date:   Sun Aug 21 14:16:23 2011 +0200

    [entropy.transceivers] EntropyFtpUriHandler: fix undefined reference

commit 1446115d33eae2e1d9f644d211596e30fcfdbe94
Author: Fabio Erculiani <lxnay@sabayon.org>
Date:   Sun Aug 21 14:13:56 2011 +0200

    [entropy.transceivers] implement remote copy method for all transceivers (similar to rename())

commit 9c55829d00cfe6abe3e61002751432e01cf1671f
Author: Fabio Erculiani <lxnay@sabayon.org>
Date:   Sun Aug 21 14:11:22 2011 +0200

    [entropy.transceivers] EntropyFileUriHandler.rename: fix typo

commit 9bb680ea67ea9432ca9636675c3aabd609f24726
Author: Fabio Erculiani <lxnay@sabayon.org>
Date:   Sun Aug 21 14:02:38 2011 +0200

    [activator] tweak commit message file name

commit 0210ee7b97a86dcd513c06fe288a4ba138007e7c
Author: Fabio Erculiani <lxnay@sabayon.org>
Date:   Sun Aug 21 11:13:45 2011 +0200

    [entropy.server.transceivers] code style update

commit 247e6fcc102d89158fe36aaefc9bb07cf2d7177a
Author: Fabio Erculiani <lxnay@sabayon.org>
Date:   Sun Aug 21 10:53:33 2011 +0200

    [services] call env-update with full path

commit d490a84b8429e7d712e8d395cf0ec8284e150894
Author: Sławomir Nizio <slawomir.nizio@sabayon.org>
Date:   Sat Aug 20 21:17:36 2011 +0200

    [po] update Polish translation

commit bc4e780274f6383678849de19e7022a30a04abac
Author: Sławomir Nizio <slawomir.nizio@sabayon.org>
Date:   Sat Aug 20 21:15:09 2011 +0200

    [misc] fix some typos in equo and manual pages

commit a4bda31855caa62adeac8650e27f93677e4e61c2
Author: Sławomir Nizio <slawomir.nizio@sabayon.org>
Date:   Sat Aug 20 18:14:13 2011 +0200

    [man] small correction in manual page, update POD (etc.) generator

commit 97f4a3d6f9a1ce832829c9310f6d49e30340a96a
Author: Fabio Erculiani <lxnay@sabayon.org>
Date:   Sat Aug 20 14:01:31 2011 +0200

    Tagging Entropy version 1.0_rc42

commit 81faac522bb4d6ac94118c0e3b47a1c21b6d03ec
Author: Fabio Erculiani <lxnay@sabayon.org>
Date:   Sat Aug 20 14:01:23 2011 +0200


server/reagent.py
#!/usr/bin/python2 -O
# -*- coding: utf-8 -*-
"""

    @author: Fabio Erculiani <lxnay@sabayon.org>
    @contact: lxnay@sabayon.org
    @copyright: Fabio Erculiani
    @license: GPL-2

    B{Entropy Package Manager Server}.

"""
import os, sys
sys.path.insert(0, '../libraries')
sys.path.insert(1, '../client')
sys.path.insert(2, '../server')
sys.path.insert(3, '/usr/lib/entropy/client')
sys.path.insert(4, '/usr/lib/entropy/libraries')
sys.path.insert(5, '/usr/lib/entropy/server')
from entropy.i18n import _
import entropy.tools
from entropy.output import red, print_error, print_generic, \
    is_stdout_a_tty, nocolor, etpUi, darkgreen
from entropy.const import etpConst, const_kill_threads
from entropy.server.interfaces import Server
from text_tools import print_menu

help_opts = [
    None,
    (0, " ~ reagent ~ ", 1,
        'Entropy Framework Server - (C) %s' % (entropy.tools.get_year(),) ),
    None,
    (0, _('Basic Options'), 0, None),
    None,
    (1, '--help', 2, _('this output')),
    (1, '--version', 1, _('print version')),
    (1, '--nocolor', 1, _('disable colorized output')),
    (1, '--color', 2, _('force colorized output')),
    None,
    (0, _('Application Options'), 0, None),
    None,
    (1, 'status', 3, _('show current repositories status')),
    None,
    (1, 'update', 3, _('scan the System looking for newly compiled packages')),
        (2, '--seekstore', 2, _('analyze the Entropy Store directory directly')),
        (2, '--repackage <atoms>', 1, _('repackage the specified atoms')),
        (2, '--noask', 2, _('do not ask anything except critical things')),
        (2, '--atoms <atoms>', 1, _('manage only the specified atoms')),
        (2, '--interactive', 2, _('run in interactive mode (asking things one by one)')),
    None,
    (1, 'inject <packages>', 1, _('add binary packages to repository w/o affecting scopes (multipackages)')),
    None,
    (1, 'query', 3, _('do some searches into repository databases')),
        (2, 'search', 3, _('search packages inside the default repository database')),
        (2, 'match', 3, _('match package dependency inside the default repository database')),
        (2, 'needed', 3, _('show runtime libraries needed by the provided atoms')),
        (2, 'revdeps', 2, _('show what packages depend on the provided atoms')),
        (2, 'tags', 3, _('show packages owning the specified tags')),
        (2, 'sets', 3, _('search available package sets')),
        (2, 'files', 3, _('show files owned by the provided atoms')),
        (2, 'belongs', 2, _('show from what package the provided files belong')),
        (2, 'description', 2, _('search packages by description')),
        (2, 'list', 3, _('list all the packages in the default repository')),
        (2, 'graph', 3, _('show direct depdendencies tree for provided installable atoms')),
            (3, '--complete', 1, _('include system packages, build deps and circularity information')),
        (2, 'revgraph', 2, _('show reverse depdendencies tree for provided installed atoms')),
            (3, '--complete', 1, _('include system packages, build deps and circularity information')),
        (2, '--verbose', 2, _('show more details')),
        (2, '--quiet', 2, _('print results in a scriptable way')),
    None,
    (1, 'repo', 3, _('manage a repository')),
        (2, 'init', 3, _('(re)initialize the current repository')),
        (2, 'bump', 4, _('manually force a revision bump for the current repository database')),
            (3, '--sync', 3, _('synchronize the database')),
        (2, 'flushback [branches]', 2, _('flush back old branches packages to current branch')),
        (2, 'remove', 4, _('remove the provided atoms from the current repository database')),
            (3, '--nodeps', 2, _('do not include reverse dependencies')),
        (2, 'multiremove', 3, _('remove the provided injected atoms (all if no atom specified)')),
        (2, 'create-empty-database <path>', 1, _('create an empty repository database in the provided path')),
        (2, 'switchbranch <from branch> <to branch>', 3, _('switch to the specified branch the repository')),
        (2, 'md5remote [atoms]', 2, _('verify remote integrity of the provided atoms')),
        (2, 'backup', 4, _('backup current repository database')),
        (2, 'restore', 4, _('restore a previously backed-up repository database')),
        (2, 'enable <repo>', 3, _('enable the specified repository')),
        (2, 'disable <repo>', 3, _('disable the specified repository')),
        (2, 'package-dep-check <repo> [atoms]', 0, _('check packages in repository for missing dependencies')),
        (2, 'package-dep <repo> [atoms]', 1, _('handle packages dependencies')),
        (2, 'package-tag <repo> <tag-string> [atoms]', 1, _('clone a package assigning it an arbitrary tag')),
        (2, 'package-mask <repo> [atoms]', 1, _('mask given package in given repository')),
        (2, 'package-unmask <repo> [atoms]', 1, _('unmask given packages in given repository')),
        (2, 'move <from> <to> [atoms]', 1, _('move packages from a repository to another')),
            (3, '--deps', 3, _('pulls dependencies in')),
        (2, 'copy <from> <to> [atoms]', 1, _('copy packages from a repository to another')),
            (3, '--deps', 3, _('pulls dependencies in')),
        (2, 'default <repo_id>', 2, _('set the default repository')),
    None,
    (1, 'key', 3, _('manage repository digital signatures (OpenGPG)')),
        (2, 'create [repos]', 1, _('create keypair for repositories and sign packages')),
        (2, 'delete [repos]', 1, _('delete keypair (and digital signatures) of repository')),
        (2, 'status [repos]', 1, _('show currently configured keys information for given repositories')),
        (2, 'sign [repos]', 1, _('sign (or re-sign) packages in repository using currently set keypair')),
        (2, 'import <repo_id> <privkey_path> <pubkey_path>', 1, _('import keypair, bind to given repository')),
        (2, 'export-public <repo_id> <key_path>', 1, _('export public key of given repository')),
        (2, 'export-private <repo_id> <key_path>', 1, _('export private key of given repository')),
    None,
    (1, 'spm', 3, _('source package manager functions')),
        (2, 'compile', 3, _('compilation function')),
            (3, 'categories', 2, _('compile packages belonging to the provided categories')),
                (4, '--list', 2, _('just list packages')),
                (4, '--nooldslots', 1, _('do not pull old package slots')),
            (3, 'pkgset', 3, _('compile packages in provided package set names')),
                (4, '--list', 2, _('just list packages')),
                (4, '--rebuild', 1, _('rebuild everything')),
                (4, '--dbupdate', 1, _('run database update if all went fine')),
                (4, '--dbsync', 1, _('run mirror sync if all went fine')),
        (2, 'orphans', 3, _('scan orphaned packages in SPM')),
        (2, 'new [categories]', 2, _('scan new packages available in SPM')),
    None,
    (1, 'deptest', 2, _('look for unsatisfied dependencies')),
    (1, 'libtest', 2, _('look for missing libraries (scan system)')),
        (2, '--dump', 2, _('dump results to files')),
    (1, 'linktest [excluded libraries]', 1, _('look for missing libraries (scan repository metadata)')),
    (1, 'pkgtest', 2, _('verify the integrity of local package files')),
    None,
    (1, 'cleanup', 2, _('remove downloaded packages and clean temp. directories)')),
    None,
]

def handle_exception(exc_class, exc_instance, exc_tb):

    # restore original exception handler, to avoid loops
    uninstall_exception_handler()
    entropy.tools.kill_threads()

    if exc_class is KeyboardInterrupt:
        raise SystemExit(1)

    # always slap exception data (including stack content)
    entropy.tools.print_exception(tb_data = exc_tb)

    raise exc_instance

def install_exception_handler():
    sys.excepthook = handle_exception

def uninstall_exception_handler():
    sys.excepthook = sys.__excepthook__

def get_entropy_server():
    return Server(community_repo = etpConst['community']['mode'])

options = sys.argv[1:]

# print version
if ("--version" in options) or ("-V" in options):
    print_generic("reagent: "+etpConst['entropyversion'])
    raise SystemExit(0)

import re
opt_r = re.compile("^(\\-)([a-z]+)$")
for n in range(len(options)):
    if opt_r.match(options[n]):
        x = options[n]
        del options[n]
        options.extend(["-%s" % (d,) for d in x[1:]])

# preliminary options parsing
_options = []
force_color = False
for opt in options:
    if opt in ["--nocolor", "-N"]:
        nocolor()
    elif opt in ["--color", "-C"]:
        force_color = True
    elif opt in ["--quiet", "-q"]:
        etpUi['quiet'] = True
    elif opt in ["--verbose", "-v"]:
        etpUi['verbose'] = True
    elif opt in ["--ask", "-a"]:
        etpUi['ask'] = True
    elif opt in ["--pretend", "-p"]:
        etpUi['pretend'] = True
    else:
        _options.append(opt)
options = _options

# Check if we need to disable colors
if (not force_color) and (not is_stdout_a_tty()):
    nocolor()

# print help
if not options or ("--help" in options) or ("-h" in options):
    print_menu(help_opts)
    if len(options) < 1:
        print_error("not enough parameters")
    raise SystemExit(1)

rc = -10
if not entropy.tools.is_root():
    print_error("you must be root in order to run "+sys.argv[0])

install_exception_handler()
main_cmd = options.pop(0)

if main_cmd == "update":
    import server_reagent
    rc = server_reagent.update(options)

elif main_cmd == "inject":
    import server_reagent
    rc = server_reagent.inject(options)

elif main_cmd == "query":
    import server_query
    rc = server_query.query(options)

elif main_cmd == "repo":
    if "switchbranch" in options:
        etpUi['warn'] = False
    import server_reagent
    rc = server_reagent.repositories(options)

elif main_cmd == "status":
    import server_reagent
    server_reagent.status()
    rc = 0

elif main_cmd == "key":
    import server_key
    rc = server_key.key(options)

elif main_cmd == "deptest":
    acquired = False
    server = None
    rc = 1
    try:
        server = get_entropy_server()
        acquired = entropy.tools.acquire_entropy_locks(server)
        if not acquired:
            print_error(darkgreen(_("Another Entropy is currently running.")))
        else:
            server.extended_dependencies_test(server.repositories())
            rc = 0
    finally:
        if server is not None:
            if acquired:
                entropy.tools.release_entropy_locks(server)
            server.shutdown()

elif main_cmd == "pkgtest":
    acquired = False
    server = None
    rc = 1
    try:
        server = get_entropy_server()
        acquired = entropy.tools.acquire_entropy_locks(server)
        if not acquired:
            print_error(darkgreen(_("Another Entropy is currently running.")))
        else:
            server._verify_local_packages(server.repository(), 
                [], ask = etpUi['ask'])
            rc = 0
    finally:
        if server is not None:
            if acquired:
                entropy.tools.release_entropy_locks(server)
            server.shutdown()

elif main_cmd == "libtest":
    dump = "--dump" in options
    acquired = False
    server = None
    rc = 1
    try:
        server = get_entropy_server()
        acquired = entropy.tools.acquire_entropy_locks(server)
        if not acquired:
            print_error(darkgreen(_("Another Entropy is currently running.")))
        else:
            rc = server.test_shared_objects(server.repository(),
                dump_results_to_file = dump)
    finally:
        if server is not None:
            if acquired:
                entropy.tools.release_entropy_locks(server)
            server.shutdown()

elif main_cmd == "linktest":
    acquired = False
    server = None
    rc = 1
    try:
        server = get_entropy_server()
        acquired = entropy.tools.acquire_entropy_locks(server)
        if not acquired:
            print_error(darkgreen(_("Another Entropy is currently running.")))
        else:
            srv_set = server.Settings()[Server.SYSTEM_SETTINGS_PLG_ID]['server']
            base_repository_id = srv_set['base_repository_id']
            qa = server.QA()
            for repository_id in server.repositories():
                repo = server.open_repository(repository_id)
                found_something = qa.test_missing_runtime_libraries(
                    server,
                    [(x, repository_id) for x in repo.listAllPackageIds()],
                    base_repository_id = base_repository_id,
                    excluded_libraries = options)
                if not found_something:
                    rc = 0
    finally:
        if server is not None:
            if acquired:
                entropy.tools.release_entropy_locks(server)
            server.shutdown()

# cleanup
elif main_cmd == "cleanup":
    import text_tools
    acquired = False
    server = None
    rc = 1
    try:
        server = get_entropy_server()
        acquired = entropy.tools.acquire_entropy_locks(server)
        if not acquired:
            print_error(darkgreen(_("Another Entropy is currently running.")))
        else:
            rc = text_tools.cleanup([etpConst['packagestmpdir'],
                etpConst['logdir']])
    finally:
        if server is not None:
            if acquired:
                entropy.tools.release_entropy_locks(server)
            server.shutdown()

elif main_cmd == "spm":
    import server_reagent
    rc = server_reagent.spm(options)

if rc == -10:
    print_menu(help_opts)
    print_error(red(" %s." % (_("Wrong parameters"),) ))
    rc = 10

uninstall_exception_handler()
entropy.tools.kill_threads()
raise SystemExit(rc)
